src/tools/kdb/main.c

```diff
@@ -0,0 +1,322 @@
+		if (tmp != NULL)
+		{
+			const char * number = keyString (keyGetMeta (errorKey, buff));
+			fprintf (stderr, "WARNING [%s]: %s\n", number, reason);
+		}
+		else
+		{
+			fprintf (stderr, "WARNING: %s\n", reason);
+		}
+	}
+}
+
+void printVersion (void)
+{
+	KeySet * versions = ksNew (0, KS_END);
+	Key * parentKey = keyNew ("system:/elektra/version", KEY_END);
+	KDB * kdb = kdbOpen (NULL, parentKey);
+	kdbGet (kdb, versions, parentKey);
+
+	Key * kdb_version = ksLookupByName (versions, "system:/elektra/version/constants/KDB_VERSION", 0);
+	if (!kdb_version)
+	{
+		fprintf (stderr, "Could not lookup KDB_VERSION key\n");
+	}
+	else
+	{
+		printf ("KDB_VERSION: %s\n", keyString (kdb_version));
+	}
+
+	Key * so_version = ksLookupByName (versions, "system:/elektra/version/constants/SO_VERSION", 0);
+	if (!so_version)
+	{
+		fprintf (stderr, "Could not lookup SO_VERSION key\n");
+	}
+	else
+	{
+		printf ("SO_VERSION: %s\n", keyString (so_version));
+	}
+
+	keyDel (parentKey);
+	ksDel (versions);
+	kdbClose (kdb, NULL);
+}
+
+int main (int argc, char ** argv)
+{
+	if (argc == 2 && (elektraStrCmp (argv[1], "--version") == 0 || elektraStrCmp (argv[1], "-V") == 0))
+	{
+		printVersion ();
+		return 0;
+	}
+	if (argc == 1)
+	{
+		printf ("kdb is a program to manage Elektra's key database. For more information use --help.\n\n");
+		return 0;
+	}
+
+	// commands where the cpp spec is not representable with libelektra opts, optional indexed args
+	for (unsigned long i = 0; i < sizeof (cppSubcommandsNoExec) / sizeof (cppCommand); ++i)
+	{
+		if (cppSubcommandsNoExec[i].exec == NULL && elektraStrCmp (cppSubcommandsNoExec[i].name, argv[1]) == 0)
+		{
+			return cpp_main (argc, argv);
+		}
+	}
+
+
+	Key * parentKey = keyNew (CLI_SPEC_KEY, KEY_END);
+	KeySet * options = ksNew (1,
+				  keyNew (CLI_SPEC_KEY, KEY_META, "command", "", KEY_META, "description",
+					  "kdb is a program to manage Elektra's key database.", KEY_END),
+				  KS_END);
+
+	// external programs with spec in spec:/sw/elektra/kdb/#0/current/<cmd>
+	KeySet * externalBinaries = ksNew (10, KS_END);
+	loadExternalSpec (options, externalBinaries, parentKey);
+
+	// C spec
+	for (unsigned long i = 0; i < sizeof (subcommands) / sizeof (subcommands[0]); ++i)
+	{
+		subcommands[i].addSpec (options);
+	}
+
+	// C++ spec
+	for (unsigned long i = 0; i < sizeof (cppSubcommands) / sizeof (cppSubcommands[0]); ++i)
+	{
+		cppSubcommands[i].addSpec (options);
+	}
+
+	int result = elektraGetOpts (options, argc, (const char **) argv, (const char **) environ, parentKey);
+	if (result == 1)
+	{ // --help
+		const char * helpMessage = elektraGetOptsHelpMessage (parentKey, NULL, NULL);
+		fprintf (stderr, "%s\n", helpMessage);
+		printWarnings (parentKey);
+		keyDel (parentKey);
+		ksDel (options);
+		ksDel (externalBinaries);
+		return 0;
+	}
+	if (result == -1)
+	{ // error
+		const char * errorMessage = GET_ERR (parentKey);
+		if (elektraStrNCmp (errorMessage, "Unknown sub-command:", 20) == 0)
```

Comment:
A bit hacky, but I guess there isn't really another option. Let's just hope, that these strings here will never get localized ...

Reply:
there is already an issue for that: #4953